include/ck_tile/ops/epilogue/cshuffle_epilogue.hpp

```diff
@@ -361,5 +362,132 @@ struct CShuffleEpilogue
+            TileDistributionEncodingPattern2D<kBlockSize,
+                                              MPerIterationShuffle,
+                                              NPerIterationShuffle,
+                                              GetVectorSizeC(),
+                                              tile_distribution_pattern::thread_raked,
+                                              Problem::kNumWaveGroups>;
+        constexpr auto dram_tile_distribution = TileEncodingPattern::Make2DStaticTileDistribution();
+
+        auto d_dram_windows = generate_tuple(
+            [&](auto idx) {
+                return make_tile_window(ds_dram_windows[idx], dram_tile_distribution);
+            },
+            number<NumDTensor>{});
+
+        constexpr auto c_warp_y_lengths =
+            to_sequence(CWarpDstr{}.get_ys_to_d_descriptor().get_lengths());
+        constexpr auto c_warp_y_index_zeros = uniform_sequence_gen_t<CWarpDstr::NDimY, 0>{};
+        // TODO: try to get rid off these here, and get them from elsewhere
+        constexpr int kM2 = 4;               // Val
+        constexpr int kM1 = (64 / NPerXdl); // Thr
+        constexpr int kM0 = MPerXdl / kM1;  // Val
+        const index_t iMWarp = get_warp_id() / NWave;
+        const index_t iNWarp = get_warp_id() - iMWarp * NWave;
+        const index_t iMLane = get_lane_id() / NPerXdl;
+        const index_t iNLane = get_lane_id() % NPerXdl;
+        static_for<0, num_access, 1>{}([&](auto iAccess) {
+            block_sync_lds();
+            constexpr auto idx_y_start = SFC::get_index(iAccess);
+
+            constexpr auto mIter = number<idx_y_start.at(number<0>{}) / (MPerIterationShuffle)>{};
+            constexpr auto nIter = number<idx_y_start.at(number<1>{}) / (NPerIterationShuffle)>{};
+
+            lds_tile.get_thread_buffer() = o_acc_tile.get_y_sliced_thread_data(
+                merge_sequences(
+                    sequence<mIter * NumMXdlPerWavePerShuffle, nIter * NumNXdlPerWavePerShuffle>{},
+                    c_warp_y_index_zeros),
+                merge_sequences(sequence<NumMXdlPerWavePerShuffle, NumNXdlPerWavePerShuffle>{},
+                                c_warp_y_lengths));
+
+            // TODO: do this properly without so much direct indexing by wrapping scale_m/n
+            // via the cktile constructs (tensor view/window etc.) instead of using a raw pointer
+            // TODO: verify if this is correct layout to use
+            auto m1       = iMLane;
+            float scale_B = scale_n[nIter * NPerXdl * NWave + iNWarp * NPerXdl + iNLane];
+            static_for<0, kM0, 1>{}([&](auto m0) {
+                static_for<0, kM2, 1>{}([&](auto m2) {
+                    float scale_A = scale_m[mIter * MPerXdl * MWave + iMWarp * MPerXdl +
+                                            m0 * kM1 * kM2 + m1 * kM2 + m2];
+                    lds_tile.get_thread_buffer()[m0 * kM2 + m2] *= scale_A * scale_B;
+                });
+            });
```

Comment:
Maybe is good to put all of these scaling features in a function:

```suggestion
            // Apply quantization scaling if provided
            if constexpr(!std::is_same_v<ScaleM, void> && !std::is_same_v<ScaleN, void>) {
                apply_quantization_scaling(...);
            }
```